### ISSUE TITLE
Comments inside record field types

### ISSUE THREAD

**Issue by ghc-mirror** [haskell/haddock#82]
**Original reporter**: _naesten@<hidden>_

for some reason,

``` haskell
newtype P a = P { runP ::
                        String          -- ^ Input string
                     -> Int             -- ^ Current column
                     -> Int             -- ^ Current line
                     -> SrcLoc          -- ^ Location of last token read
                     -> ParseState      -- ^ Layout info.
                     -> ParseMode       -- ^ Parse parameters
                     -> ParseStatus a
                }
```

fails to parse, but

``` haskell
newtype P a = P { runP ::
                        String          -- Input string
                     -> Int             -- Current column
                     -> Int             -- Current line
                     -> SrcLoc          -- Location of last token read
                     -> ParseState      -- Layout info.
                     -> ParseMode       -- Parse parameters
                     -> ParseStatus a
                }
```

parses fine.

**Comment by sjakobi** [comment]
Duplicate of #23?

**Comment by harpocrates** [comment]
> Duplicate of #23?

Definitely not - this is much simpler! In fact, I think this is quite achievable even today. We just need to tweak the ghc grammar to replace on `sigtype` with a `sigtypedoc`. The ambiguity pointed out above is the main problem.

> This is just not supported by the current grammar. It would be nice to support it, but I wonder how, since it is possible to comment record fields using the -- ^ notation which would clash with commenting the result type.

I think it would be quite reasonable to take the same approach to resolving this ambiguity as we take for constructor arguments.

I might take a stab at this once the Hi Haddock changes land in GHC.

**Comment by schell** [comment]
Gently pinging this as I ran into this today. I'd like to document my record like this:
```haskell
data MyRecord m
  = MyRecord
  { -- | A function that does stuff. 
    myFunction 
      :: V2 Int 
      -- ^ The position  
      -> m SomeType 
      -- ^ The return type
  }
```

I don't have all the context, but if a comment is found inside the type of a record field, then I think it's safe t0 assume that the author's intent is to document those field's types and the return value instead of the field itself. The field itself can still be documented using `-- |`. This is how GADTs are documented so I was surprised to find that records don't work this way. 

Anyway - thanks for these great tools!

**Comment by harpocrates** [comment]
Thanks for the ping - this is still very much on my radar. Hi Haddock is not yet done, though it hopefully will be come end of next month! Once that wraps up, there will be more time to work on this sort of enhancement.

**Comment by sirlensalot** [comment]
Is this something a noob contrib like myself can help with? Just lost significant time chasing this problem down. FWIW it's actually pretty hard to find out about this limitation, it might be nice to put a gotcha like this in the README or somewhere easy to find.

**Comment by Kleidukos** [comment]
> Is this something a noob contrib like myself can help with? Just lost significant time chasing this problem down. FWIW it's actually pretty hard to find out about this limitation, it might be nice to put a gotcha like this in the README or somewhere easy to find.

A PR adding this limitation for the ReST docs would be welcome. Tag me in it so that I can provide a timely review.
